### ISSUE TITLE
Add the ability to copy messages in Expensify.cash

### ISSUE THREAD

**Issue by roryabraham** [Expensify/Expensify.cash#1778]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/master/CONTRIBUTING.md) for onboarding!
___

**Internal issue:** https://github.com/Expensify/Expensify/issues/147479
**Internal Upwork Posting:** https://www.upwork.com/ab/applicants/1371557081012895744/job-details
**Public Upwork Posting:**  https://www.upwork.com/jobs/~013badfb341950b5a6
___

## Expected Result:

The "Copy to Clipboard" button in the `ReportActionContextMenu` should copy the text context of a report action / message to the clipboard.

## Actual Result:
The "Copy to Clipboard" button in the `ReportActionContextMenu` currently doesn't do anything:

![image](https://user-images.githubusercontent.com/47436092/111212780-45088000-858d-11eb-8167-6a8a40799749.png)

---

![image](https://user-images.githubusercontent.com/47436092/111212832-5782b980-858d-11eb-9868-c4755d1802a9.png)

---

![image](https://user-images.githubusercontent.com/47436092/111212862-610c2180-858d-11eb-88b3-ed5ca449acb4.png)

## Action Performed:
1. Temporarily edit [this function](https://github.com/Expensify/Expensify.cash/blob/54b301738c5606e3c5b2e4614a9f05d78e77c992/src/pages/home/report/ReportActionItem.js#L67-L70) to always return `true` so that the `ReportActionContextMenu` will show (see screenshots above).
2. Hover, right-click, or longPress a message in a report (called a reportAction), then click `Copy to Clipboard`

## Workaround:
None.

## Platform:
All platforms ->

 Web
 iOS
 Android
 Desktop App
 Mobile Web
 
**Notes/Photos/Videos:** 
- A good clipboard library for React Native is [@react-native-community/clipboard](https://www.npmjs.com/package/@react-native-community/clipboard), so your solution should probably leverage that.
- Your solution must work on all platforms.
- A solution should copy only the plain-text version of a message, not HTML the html.
- BONUS: Make it also work to copy images 😉

**Comment by MazenRefaat** [comment]
I'm trying to run it locally, don't know how to get API to run

**Comment by SameeraMadushan** [comment]
Working on proposal.

**Comment by parasharrajat** [comment]
Hi, I would like to work on this issue. It has scope for me to learn the Native Modules.

## Proposal
1. Add an onPress property to the `CONTEXT_ACTIONS` in the `ReportActionContextMenu.js` which will be called on the OnPress action something like.
```javascript
 {
        text: 'Copy to Clipboard',
        icon: Clipboard,
        onPress: (action) => {
            console.debug(action);
          if(image message) {
              ClipboardAPI.setImage(action.message[0].html);
         } else{
            ClipboardAPI.setString(action.message[0].text);
        },
    },
    ...............
 <ReportActionContextMenuItem
                    icon={contextAction.icon}
                    text={contextAction.text}
                    isMini={props.isMini}
                    onPress={() => contextAction.onPress(props.action)}
                    key={contextAction.text}
                />
``` 
2. We will send the action prop from the Parent Context.
3. Now as mentioned on the issue, we can utilize the `import { Clipboard } from 'react-native;` as we are using React Native v0.63.x which has his API. or we can use the `@react-native-community/clipboard ` package which has the same API.
4. If we are using the  `@react-native-community/clipboard ` lib we have to create a wrapper function to resolve the  `Clipboard` API as
 `@react-native-community/clipboard `exports the Clipboard as default but react-native-web export it as named export.
 5. Above solution will provide copying text strings for **All the Platforms.**. 
 
 ### Images
 Copying images is not supported by either react-native or the package. 
 
 #### For Browser-based platforms
 1. `React native web`  
  https://github.com/necolas/react-native-web/blob/66d01734ce3ffcfea61e05aa1b45015658b3f2af/packages/react-native-web/src/exports/Clipboard/index.js#L33-L46
 As you see that they are using a `span` and setting the `textcontent` which escapes the Html which is needed to copy the images on the Browsers.

So we can extend the lib and create a Method same as the `react-native-web` `setString` called `setImage` which uses a `contenteditable` `div` and we set the innerHTML with the HTML format of the action for image messages and then the copy will include the following.
```
type: text/html
            <img src="https://res.cloudinary.com/practicaldev/image/fetch/s---V8aGC_g--/c_limit%2Cf_auto%2Cfl_progressive%2Cq_auto%2Cw_880/https://dev-to-uploads.s3.amazonaws.com/i/zzfake78sf0bqwswcfn4.png" alt="Copy Image to Clipboard via Javascript.">
```
#### For Native Platform IOS
1. We can use the native Bridging as IOS has https://developer.apple.com/documentation/uikit/uipasteboard.
This is exactly what we need to do for copying images in IOS. https://stackoverflow.com/a/44423850/7462668.

#### For Native Platform Android
1. This one is kind of tricky. We need native bridging.
https://developer.android.com/guide/topics/text/copy-paste.html#Copying This article suggests using the URI for the complex data types.
 I am naive to Native Modules but I will give my best.

--- 
 So In short
 > **Create Clipboard lib which exports a class Clipboard that extends the Clipboard from the `@react-native-community/clipboard` or `react-native` and adds `setImage` method which will be implemented for native and browser-based platforms as per above instructions.**

**Comment by roryabraham** [comment]
@parasharrajat Your proposal looks pretty good, but I do have some comments:

1) I think we'll want to use `@react-native-community/clipboard` component because the one that ships with `RN` is deprecated:

![image](https://user-images.githubusercontent.com/47436092/112038434-0f6d1500-8b00-11eb-9846-19ddd4f3e7ab.png)

2)  > We will send the action prop from the Parent Context.

Let's actually just replace the `reportActionID` prop in `ReportActionContextMenu` with `reportAction`, and pass the whole action instead of just the ID.

3) > Do we need a selection copy or just the whole message as per the ContextMenu I think full message?

Yep, this feature should just copy the whole message. On web/desktop, highlighting text and using `CMD+C` should still work, but on native platforms there won't really be a way to copy a partial message, because the `longPress` action that would typically allow one to highlight and select text will instead open the custom context menu.

4) > Do we need to remove the HTML tags from it like we have done for the last message preview in the sidebar

Yes, content copied to clipboard should have all HTML tags stripped first.

---

Now, the **most important caveat** of this proposal's acceptance is that I think for now we should just focus on getting text copied successfully. We can worry about images in a separate issue. We really don't want to manage or maintain any native code in Expensify.cash if we can avoid it, so the only really acceptable solution there in my opinion would to be to:

1) Fork `@react-native-community/clipboard`, and implement image-copying in the fork.
2) Point Expensify.cash at the fork and implement the new image copying functionality in Expensify.cash
3) Submit a PR for [this issue](https://github.com/react-native-clipboard/clipboard/issues/6), get it merged.
4) Once that's merge, update Expensify.cash to point at the correctly updated version of `react-native-clipboard` instead of the fork.

AFAIK this isn't really something we've done yet.

cc @marcaaron @tgolen in case you guys have thoughts.

**Comment by roryabraham** [comment]
Also @parasharrajat, after the text is copied to the clipboard, can we change the clipboard icon to a green checkmark and have the tooltip text display `Copied!` until the menu is closed and reopened. For a demo of a similar experience, try copying the branch name of the head branch of a pull request in Github:

![image](https://user-images.githubusercontent.com/47436092/112041985-3b8a9500-8b04-11eb-811c-daaf06f77309.png)

**Comment by parasharrajat** [comment]
@roryabraham Thanks for clarifying the doubts.
I also agree that we should move the images part to the new issue as this involves a little more and especially the native part. So 
I will work over the text part for now.

About the tick, surely we can do that. But where do you want to show that? On the context menu or hover menu.

**Comment by tgolen** [comment]
@roryabraham I like your proposal for the image support. I think we should contact the contributors directly and offer to pay them to do that issue, rather than just waiting for them to prioritize it on their own.

**Comment by roryabraham** [comment]
Well, just to clarify the "context menu" and "hover menu" are just different manifestations of the `ReportActionContextMenu`, and they should never show at the same time. A similar effect with the checkmark should be implemented on both, ideally.

**Comment by mallenexpensify** [comment]
@roryabraham Assigned to you since you'll be the one reviewing proposals

**Comment by roryabraham** [comment]
I've already chosen a contributor, and he has a WIP PR up

**Comment by roryabraham** [comment]
Sorry for the confusion, but per our documented process:

> The engineer assigned to the Expensify/Expensify Cash issue should review incoming proposals and give the go ahead to a contributor to start working on a PR and should assign the Expensify/Expensify Cash issue to the contributor.

So I'm actually re-assigning this to @parasharrajat

**Comment by kaushiktd** [comment]
Hi,

This is Kaushik here and I would like to suggest a simple and effective explanation for this,

  1.  Add an onPress property to call clipboard function
        ClipboardAPI.setString(action.message[0].text);
